rpc/server.go

```diff
@@ -39,178 +40,84 @@ func NewRpcServer(nitroClient *nitro.Client, chainId *big.Int, logger zerolog.Lo
-	handleError(err)
+	if err != nil {
+		panic(err)
+	}
 
 	rs := &RpcServer{natstrans.NewNatsConnection(nc), ns, nitroClient, chainId, logger}
-	rs.registerHandlers()
+	err = rs.registerHandlers()
+	if err != nil {
+		panic(err)
+	}
+
 	return rs
 }
 
-// registerHandlers registers the handlers for the rpc server
-func (rs *RpcServer) registerHandlers() {
-	err := rs.connection.Subscribe(serde.DirectFundRequestMethod, func(data []byte) []byte {
-
+func subcribeToRequest[T serde.RequestPayload, U serde.ResponsePayload](rs *RpcServer, method serde.RequestMethod, processPayload func(T) U) error {
+	return rs.connection.Subscribe(method, func(data []byte) []byte {
 		rs.logger.Trace().Msgf("Rpc server received request: %+v", data)
-
-		rpcRequest := serde.JsonRpcRequest[directfund.ObjectiveRequest]{}
+		rpcRequest := serde.JsonRpcRequest[T]{}
 		err := json.Unmarshal(data, &rpcRequest)
 		if err != nil {
 			panic("could not unmarshal direct fund objective request")
 		}
+		obj := rpcRequest.Params
+		objResponse := processPayload(obj)
 
-		// todo: objective request is redefined so that it has a valid objectiveStarted channel.
-		// 	Should find a better way to accomplish this.
-		objectiveRequestWithChan := directfund.NewObjectiveRequest(
-			rpcRequest.Params.CounterParty,
-			rpcRequest.Params.ChallengeDuration,
-			rpcRequest.Params.Outcome,
-			rpcRequest.Params.Nonce,
-			rpcRequest.Params.AppDefinition,
-		)
-
-		rs.client.IncomingObjectiveRequests() <- objectiveRequestWithChan
-
-		objRes := rpcRequest.Params.Response(*rs.client.Address, rs.chainId)
-		msg := serde.NewJsonRpcResponse(rpcRequest.Id, objRes)
+		msg := serde.NewJsonRpcResponse(rpcRequest.Id, objResponse)
 		messageData, err := json.Marshal(msg)
 		if err != nil {
 			panic("Could not marshal direct fund response message")
 		}
 
 		return messageData
 	})
-	if err != nil {
-		panic(err)
-	}
-
-	err = rs.connection.Subscribe(serde.DirectDefundRequestMethod, func(data []byte) []byte {
-		rs.logger.Trace().Msgf("Rpc server received request: %+v", data)
-
-		rpcRequest := serde.JsonRpcRequest[directdefund.ObjectiveRequest]{}
-		err := json.Unmarshal(data, &rpcRequest)
-		if err != nil {
-			panic("could not unmarshal direct fund objective request")
-		}
-
-		// todo: objective request is redefined so that it has a valid objectiveStarted channel.
-		// 	Should find a better way to accomplish this.
-		objectiveRequestWithChan := directdefund.NewObjectiveRequest(
-			rpcRequest.Params.ChannelId,
-		)
-
-		rs.client.IncomingObjectiveRequests() <- objectiveRequestWithChan
-
-		objRes := rpcRequest.Params.Id(*rs.client.Address, rs.chainId)
-		msg := serde.NewJsonRpcResponse(rpcRequest.Id, objRes)
-		messageData, err := json.Marshal(msg)
-		if err != nil {
-			panic("Could not marshal direct fund response message")
-		}
+}
 
-		return messageData
+// registerHandlers registers the handlers for the rpc server
+func (rs *RpcServer) registerHandlers() error {
```

Comment:
This is now a nice, well-contained place to add new "routes" from the RPC API to the "internal" API 🚀